server/src/main/java/org/apache/iotdb/db/engine/compaction/task/CompactionRecoverTask.java

```diff
@@ -20,102 +20,483 @@
+          TsFileIdentifier sourceFileIdentifier = sourceFileIdentifiers.get(i);
+          if (sourceFileIdentifier.isSequence()) {
+            File tmpTargetFile = targetFileIdentifiers.get(i).getFileFromDataDirs();
+            File targetFile = null;
+
+            // move tmp target file to target file if not exist
+            if (tmpTargetFile != null) {
+              // move tmp target file to target file
+              String sourceFilePath =
+                  tmpTargetFile
+                      .getPath()
+                      .replace(
+                          TsFileConstant.TSFILE_SUFFIX
+                              + IoTDBConstant.CROSS_COMPACTION_TMP_FILE_SUFFIX_FROM_OLD,
+                          TsFileConstant.TSFILE_SUFFIX);
+              targetFile = TsFileNameGenerator.increaseCrossCompactionCnt(new File(sourceFilePath));
+              FSFactoryProducer.getFSFactory().moveFile(tmpTargetFile, targetFile);
+            } else {
+              // target file must exist
+              File file =
+                  TsFileNameGenerator.increaseCrossCompactionCnt(
+                      new File(
+                          targetFileIdentifiers
+                              .get(i)
+                              .getFilePath()
+                              .replace(
+                                  TsFileConstant.TSFILE_SUFFIX
+                                      + IoTDBConstant.CROSS_COMPACTION_TMP_FILE_SUFFIX_FROM_OLD,
+                                  TsFileConstant.TSFILE_SUFFIX)));
+
+              targetFile = getFileFromDataDirs(file.getPath());
+            }
+            if (targetFile == null) {
+              LOGGER.error(
+                  "{} [Compaction][Recover] target file of source seq file {} does not exist (<0.13).",
+                  fullStorageGroupName,
+                  sourceFileIdentifier.getFilePath());
+              return false;
+            }
+
+            // serialize target resource file if not exist
+            TsFileResource targetResource = new TsFileResource(targetFile);
+            if (!targetResource.resourceFileExists()) {
+              try (TsFileSequenceReader reader =
+                  new TsFileSequenceReader(targetFile.getAbsolutePath())) {
+                FileLoaderUtils.updateTsFileResource(reader, targetResource);
+              }
+              targetResource.serialize();
+            }
+
+            targetFileResources.add(targetResource);
+
+            // append compaction modifications to target mods file and delete compaction mods file
+            if (compactionModsFileFromOld.exists()) {
+              ModificationFile compactionModsFile =
+                  new ModificationFile(compactionModsFileFromOld.getPath());
+              appendCompactionModificationsBefore013(targetResource, compactionModsFile);
+            }
+
+            // delete tmp target file
+            if (tmpTargetFile != null) {
+              tmpTargetFile.delete();
```

Comment:
check the return value of all `delete` function, or use `FileUtils.delete` which throws exception if the deletion is failed. Here we can simply new a `TsFileResource` instance, and use the `remove` method to delete file, resourace file and modification.

Reply:
Resolved. And it is better to check and delete the source file, resource file and mods file separately rather than using `TsFileResource.remove()` , because the tsfile may be deleted, and the resource file and mods file may have not been deleted.

Reply:
Get it